x-pack/plugins/proxy/server/cluster_doc.ts

```diff
@@ -0,0 +1,318 @@
+    const table = await this.getRoutingTable();
+    return table[resource];
+  }
+
+  public async getRoutingTable() {
+    const client = await this.getESClient();
+    const params = {
+      index: this.proxyIndex,
+      type: this.docType,
+      id: this.routingDoc,
+      source: true,
+    };
+    const res = await client.callAsInternalUser('get', params);
+    return res._source as RoutingTable;
+  }
+
+  public async assignResource(resource: string, type: string, state: RouteState, node?: string) {
+    const table = await this.getRoutingTable();
+    if (table[resource]) {
+      throw new Error(`${resource} already exists on ${table[resource].node}`);
+    }
+    const body = {
+      [resource]: {
+        type,
+        state,
+        node: node || this.nodeName,
+      },
+    };
+
+    const client = await this.getESClient();
+    const params = {
+      index: this.proxyIndex,
+      type: this.docType,
+      id: this.routingDoc,
+      body,
+      retryOnConflict: this.maxRetry,
+    };
+    await client.callAsInternalUser('update', params);
+  }
+
+  public async unassignResource(resource: string) {
+    const client = await this.getESClient();
+    const body = {
+      script: unassignResource,
+      params: {
+        resource,
+      },
+    };
+    const params = {
+      index: this.proxyIndex,
+      type: this.docType,
+      id: this.routingDoc,
+      body,
+      retryOnConflict: this.maxRetry,
+    };
+    await client.callAsInternalUser('update', params);
+  }
+
+  private async getESClient(): Promise<ClusterClient> {
+    if (!this.elasticsearch$) {
+      const err = Boom.boomify(new Error('You must call setup first'), { statusCode: 412 });
+      throw err;
+    }
+    const client = await this.elasticsearch$.pipe(first()).toPromise();
+    return client;
+  }
+
+  private setConfig(config: ProxyPluginType) {
+    const update = config.updateInterval + randomInt(this.minUpdateShuffle, this.maxUpdateShuffle);
+    const timeout =
+      config.timeoutThreshold + randomInt(this.minUpdateShuffle, this.maxUpdateShuffle);
+    this.updateInterval = update;
+    this.timeoutThreshold = timeout;
+    this.maxRetry = config.maxRetry;
+  }
+
+  private setTimer() {
+    if (this.updateTimer) return;
+    this.log.debug('Set timer to updateNodeMap');
+    this.updateTimer = setTimeout(() => {
+      this.log.debug('Updating heartbeat');
+      this.mainLoop();
+    }, this.updateInterval);
+  }
+
+  public async getHeartbeats() {
+    const client = await this.getESClient();
+    const params = {
+      id: this.heartbeatDoc,
+      type: this.docType,
+      index: this.proxyIndex,
+      _source: true,
+    };
+    const reply = await client.callAsInternalUser('get', params);
+    return reply._source as NodeList;
+  }
+
+  /**
+   * Node heartbeats are monotonically increasing integers
+   * @param remove [boolean] should this node be deleted
+   */
+  public async updateHeartbeat(remove: boolean = false) {
```

Comment:
shouldn't update/delete be separate methods? if we add more parameters that specific for either operation we will have to workaround this